[Small Feature] Point Set Clustering

sgiraudot opened the pull request:
## Rationale

Identifying clusters of locally connected points in a point set is a pretty standard procedure. Although the algorithm is pretty simple and straightforward, CGAL does not provide any API to such an algorithm, which seems like a missing feature.

(In my personal experience, I used several times this algorithm and kept copy-pasting the code, so I figured it belonged in the _Point Set Processing_ package.)

This small features adds a clustering algorithm to _Point Set Processing_.

## Summary of API changes

A new function `CGAL::cluster_point_set()` is provided, along with a new named parameter `adjacencies`.

## License and copyright ownership

GPL, GeometryFactory

## CHANGES.md

(todo)

## Submission

### Version 1 (_outdated_)

* [~~__User manual__~~](https://cgal.geometryfactory.com/~sgiraudot/PSP-Cluster_point_set/Point_set_processing_3/index.html#title17)
* [~~__Reference manual__~~](https://cgal.geometryfactory.com/~sgiraudot/PSP-Cluster_point_set/Point_set_processing_3/group__PkgPointSetProcessing3Algorithms.html#ga636fa37aa7ed4ae296028d4b67470a24)

### Version 2

* [__User manual__](https://cgal.geometryfactory.com/~sgiraudot/PSP-Cluster_point_set_v2/Point_set_processing_3/index.html#title17)
* [__Reference manual__](https://cgal.geometryfactory.com/~sgiraudot/PSP-Cluster_point_set_v2/Point_set_processing_3/group__PkgPointSetProcessing3Algorithms.html#gafee41d60b5a257ae034e9157d0af8e46)

## Status

- Developed and locally tested (GNU/Linux)
- [Small Feature](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Point_Set_Clustering)  --  pre-approved Andreas Fabri 10:50, 8 April 2020 (CEST) 
- __Note:__ this features depends on https://github.com/CGAL/cgal/pull/4552, so it should be integrated afterwards (I'm not really worried, small features usually take longer to be integrated than simple PR)

mglisse commented:
Precisely because clustering is such a well-known and wide topic, it should be clear what algorithm you are using. Are you just returning the connected components of a union of balls? The generic name "clustering" may give users a lot of expectations.

sgiraudot commented:
The documentation says:

> Identifies simply connected clusters on a nearest neighbors graph. 

How can I make it clearer?

mglisse commented:
simply connected clusters -> connected components

nearest neighbor graph: that has a precise definition. Neighborhood graph is more vague. I've seen the graph with all pairs at distance less than r called the r-Rips graph, proximity graph, maybe also some name with "ball" in it, not sure about it, it probably doesn't have a canonical name. Anyway, a sentence explaining what it does cannot hurt.

afabri commented:
I suggest to add an image for a 2D point set (then it is planar which is better for html).  
I think your "neighbor" function must be symmetric as you strive for equivalence classes.  It does not seem to be well defined for functions like  "a link to the closest point"  which are not symmetric.

maxGimeno commented:
Errors on travis [here](https://travis-ci.com/github/CGAL/cgal/jobs/296517761#L3502) and [here](https://travis-ci.com/github/CGAL/cgal/jobs/296517732#L7434)

sgiraudot commented:
I uploaded a new version of the doc following the different reviews I got. Following a discussion with @afabri, I removed the parameter `K`, as using a number of nearest neighbors to create the nearest neighbor graph creates a directed graph that is not well suited for this algorithm. Now only a neighbor radius can be used, which makes the graph undirected and the algorithm better defined. I also added a figure.

MaelRL commented:
@afabri @sgiraudot bump

maxGimeno commented:
Warning : https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.1-Ic-127/Polyhedron_Demo/TestReport_lrineau_Ubuntu-Latest-CXX11.gz

maxGimeno commented:
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-130.shtml

sgiraudot commented:
Is this something I can do? Will this be fixed if I add the Property_map package to the dependency files of packages listed in:
```
Differences in Voronoi_diagram_2: Property_map are new and not committed.
 Differences in Triangulation: Property_map are new and not committed.
 Differences in TDS_3: Property_map are new and not committed.
 Differences in TDS_2: Property_map are new and not committed.
 Differences in STL_Extension: Property_map are new and not committed.
 Differences in QP_solver: Property_map are new and not committed.
 Differences in Polynomial: Property_map are new and not committed.
 Differences in Point_set_3: Circulator are new and not committed.
 Differences in Mesh_2: Property_map are new and not committed.
 Differences in Generalized_map: Property_map are new and not committed.
 Differences in Convex_hull_d: Property_map are new and not committed.
 Differences in Cone_spanners_2: Property_map are new and not committed.
 Differences in Bounding_volumes: Property_map are new and not committed.
 Differences in Boolean_set_operations_2: Property_map are new and not committed.
 Differences in Algebraic_kernel_d: Property_map are new and not committed.
```

maxGimeno commented:
@sgiraudot yes, exactly

sgiraudot commented:
I ended up moving the new functions to `Property_map` which seem cleaner and shouldn't change the dependencies.

maxGimeno commented:
Circulator is a new dep of Point_set_3, apparently.

maxGimeno commented:
warnings in the demo : 
https://cgal.geometryfactory.com/CGAL/testsuite/CGAL-5.1-Ic-137/Polyhedron_Demo/TestReport_lrineau_Ubuntu-latest-GCC6-Release.gz

maxGimeno commented:
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-139.shtml.

sloriot commented:
@sgiraudot could you please update `changes.md`?

sloriot commented on Property_map/include/CGAL/property_map.h:
```diff
@@ -576,6 +578,38 @@ make_cartesian_converter_property_map(Vpm vpm)
+template <typename Range, typename Pmap>
+CGAL::Iterator_range<typename boost::transform_iterator<CGAL::Property_map_to_unary_function<Pmap>,
+                                                        typename Range::iterator> >
+make_transform_range_from_property_map (Range& range, Pmap pmap)
```
I would rename that to `transform_range_[with/using]_property_map()`

maxGimeno commented on Polyhedron/demo/Polyhedron/Plugins/Point_set/Point_set_clustering_plugin.cpp:
```diff
@@ -0,0 +1,241 @@
+
+    std::size_t nb_clusters = *functor.result;
+
+    Scene_group_item* group;
```
warning: may be used unitialized

maxGimeno commented on Polyhedron/demo/Polyhedron/Plugins/Point_set/Point_set_clustering_plugin.cpp:
```diff
@@ -0,0 +1,241 @@
+      {
+        Scene_points_with_normal_item* new_item = new Scene_points_with_normal_item;
+        new_item->point_set()->copy_properties (*points);
+        CGAL::Random rand(i);
```
conversion from 'size_t' to 'unsigned int

maxGimeno commented on Polyhedron/demo/Polyhedron/Plugins/Point_set/Point_set_clustering_plugin.cpp:
```diff
@@ -0,0 +1,241 @@
+        Point_set::Index iidx = *(colored->point_set()->insert (points->point(idx)));
+        if (cluster_size[cluster_map[idx]] >= std::size_t(min_nb->value()))
+        {
+          CGAL::Random rand(cluster_map[idx] + 1);
```
conversion from 'size_t' to 'unsigned int

maxGimeno commented:
Successfully tested in https://cgal.geometryfactory.com/CGAL/testsuite/results-5.1-Ic-152.shtml
